src/main/java/me/kavin/piped/server/handlers/auth/UserHandlers.java

```diff
@@ -109,10 +132,277 @@ public static byte[] loginResponse(String user, String pass)
         }
     }
 
-    public static byte[] deleteUserResponse(String session, String pass) throws IOException {
+    public static HttpResponse oidcLoginRequest(OidcProvider provider, String redirectUri) throws Exception {
+        if (StringUtils.isBlank(redirectUri)) {
+            return HttpResponse.ofCode(400).withHtml("redirect is a required parameter");
+        }
+
+        URI callback = new URI(Constants.PUBLIC_URL + "/oidc/" + provider.name + "/callback");
+        CodeVerifier codeVerifier = new CodeVerifier();
+        OidcData data = new OidcData(redirectUri, codeVerifier);
+        String state = data.getState();
+
+        DatabaseHelper.setOidcData(data);
+
+        AuthenticationRequest oidcRequest = new AuthenticationRequest.Builder(
+                new ResponseType("code"),
+                new Scope("openid"),
+                provider.clientID, callback
+        )
+                .endpointURI(provider.authUri)
+                .codeChallenge(codeVerifier, CodeChallengeMethod.S256)
+                .state(new State(state))
+                .nonce(data.getOidNonce()).build();
+
+        if (redirectUri.equals(Constants.FRONTEND_URL + "/login")) {
+            return HttpResponse.redirect302(oidcRequest.toURI().toString());
+        }
+        return HttpResponse.ok200().withHtml(
+                "<!DOCTYPE html><html style=\"color-scheme: dark light;\"><body>" +
+                        "<h3>Warning:</h3> You are trying to give <pre style=\"font-size: 1.2rem;\">" +
+                        redirectUri +
```

Comment:
`redirectUri` needs to be HTML escaped or checked if it's a real URL, otherwise that would allow HTML injection.